ml-dsa/src/lib.rs

```diff
@@ -1,2 +1,686 @@
+    // Algorithm 22 pkEncode
+    pub fn encode(&self) -> EncodedVerifyingKey<P> {
+        Self::encode_internal(&self.rho, &self.t1)
+    }
+
+    /// Decode the key from an appropriately sized byte array.
+    // Algorithm 23 pkDecode
+    pub fn decode(enc: &EncodedVerifyingKey<P>) -> Self {
+        let (rho, t1_enc) = P::split_vk(enc);
+        let t1 = P::decode_t1(t1_enc);
+        Self::new(rho.clone(), t1, None, Some(enc.clone()))
+    }
+}
+
+impl<P: MlDsaParams> signature::Verifier<Signature<P>> for VerifyingKey<P> {
+    fn verify(&self, msg: &[u8], signature: &Signature<P>) -> Result<(), Error> {
+        self.verify_with_context(msg, &[], signature)
+            .then_some(())
+            .ok_or(Error::new())
+    }
+}
+
+/// `MlDsa44` is the parameter set for security category 2.
+#[derive(Default, Clone, Debug, PartialEq)]
+pub struct MlDsa44;
+
+impl ParameterSet for MlDsa44 {
+    type K = U4;
+    type L = U4;
+    type Eta = U2;
+    type Gamma1 = Shleft<U1, U17>;
+    type Gamma2 = Quot<QMinus1, U88>;
+    type TwoGamma2 = Prod<U2, Self::Gamma2>;
+    type W1Bits = Length<Diff<Quot<U88, U2>, U1>>;
+    type Lambda = U32;
+    type Omega = U80;
+    const TAU: usize = 39;
+}
+
+/// `MlDsa65` is the parameter set for security category 3.
+#[derive(Default, Clone, Debug, PartialEq)]
+pub struct MlDsa65;
+
+impl ParameterSet for MlDsa65 {
+    type K = U6;
+    type L = U5;
+    type Eta = U4;
+    type Gamma1 = Shleft<U1, U19>;
+    type Gamma2 = Quot<QMinus1, U32>;
+    type TwoGamma2 = Prod<U2, Self::Gamma2>;
+    type W1Bits = Length<Diff<Quot<U32, U2>, U1>>;
+    type Lambda = U48;
+    type Omega = U55;
+    const TAU: usize = 49;
+}
+
+/// `MlKem87` is the parameter set for security category 5.
+#[derive(Default, Clone, Debug, PartialEq)]
+pub struct MlDsa87;
+
+impl ParameterSet for MlDsa87 {
+    type K = U8;
+    type L = U7;
+    type Eta = U2;
+    type Gamma1 = Shleft<U1, U19>;
+    type Gamma2 = Quot<QMinus1, U32>;
+    type TwoGamma2 = Prod<U2, Self::Gamma2>;
+    type W1Bits = Length<Diff<Quot<U32, U2>, U1>>;
+    type Lambda = U64;
+    type Omega = U75;
+    const TAU: usize = 60;
+}
+
+/// A parameter set that knows how to generate key pairs
+pub trait KeyGen: MlDsaParams {
+    /// The type that is returned by key generation
+    type KeyPair: signature::Keypair;
+
+    /// Generate a signing key pair from the specified RNG
+    #[cfg(feature = "rand_core")]
+    fn key_gen(rng: &mut impl CryptoRngCore) -> Self::KeyPair;
+
+    /// Deterministically generate a signing key pair from the specified seed
+    // TODO(RLB): Only expose this based on a feature.
+    fn key_gen_internal(xi: &B32) -> Self::KeyPair;
+}
+
+impl<P> KeyGen for P
+where
+    P: MlDsaParams,
+{
+    type KeyPair = KeyPair<P>;
+
+    /// Generate a signing key pair from the specified RNG
+    // Algorithm 1 ML-DSA.KeyGen()
+    #[cfg(feature = "rand_core")]
+    fn key_gen(rng: &mut impl CryptoRngCore) -> KeyPair<P> {
```

Comment:
We typically use the names `generate*` or `random` for these sorts of methods. See also:

https://docs.rs/crypto-common/latest/crypto_common/trait.KeyInit.html